src/Raven.Server/Storage/Schema/Updates/Documents/50000/From42017.cs

```diff
@@ -1,16 +1,493 @@
+                                    currentPrefix = counterGroup.Id;
+                                    break;
+                                }
+                            }
+
+                            // start new batch
+                            currentDocId = counterGroup.Id;
+                            batch.Counters.Add(counterGroup);
+                        }
+
+                        // add table etag to delete-list
+                        if (toDeleteByCollection.TryGetValue(counterGroup.Collection, out var toDeleteList) == false)
+                        {
+                            toDeleteByCollection[counterGroup.Collection] = toDeleteList = new List<long>();
+                        }
+                        toDeleteList.Add(counterGroup.Etag);
+                        
+                        processedInCurrentTx++;
+                    }
+
+                    if (commit)
+                    {
+                        step.Commit();
+                        step.RenewTransactions();
+                        step.DocumentsStorage.CountersStorage = new CountersStorage(step.DocumentsStorage.DocumentDatabase, step.WriteTx);
+                        currentDocId = null;
+                        continue;
+                    }
+
+                    if (toDeleteByCollection.Count > 0)
+                    {
+                        DeleteProcessedEntries(step, toDeleteByCollection);
+                    }
+
+                    if (batch.Counters.Count > 0)
+                    {
+                        PutCounterGroups(step, batch, context);
+                    }
+
+                    done = true;
+                }
+            }
+
             return true;
         }
+
+        private void PutCounterGroups(UpdateStep step, CounterBatchUpdate batch, DocumentsOperationContext context)
+        {
+            foreach (var cg in batch.Counters)
+            {
+                PutCounters(context, step, cg.Id, cg.ChangeVector, cg.Values, cg.Collection);
+            }
+
+            batch.Clear();
+        }
+
+        private static void DeleteProcessedEntries(UpdateStep step, Dictionary<string, List<long>> toDeleteByCollection)
+        {
+            foreach (var toDeleteForCollection in toDeleteByCollection)
+            {
+                var collection = new CollectionName(toDeleteForCollection.Key);
+                var table = step.WriteTx.OpenTable(CountersSchema, collection.GetTableName(CollectionTableType.CounterGroups));
+
+                foreach (var etag in toDeleteForCollection.Value)
+                {
+                    table.DeleteByIndex(CountersSchema.FixedSizeIndexes[CollectionCountersEtagsSlice], etag);
+                }
+            }
+            
+            toDeleteByCollection.Clear();
+        }
+
+        private static IEnumerable<CounterReplicationItem> GetCounters(Table table, DocumentsOperationContext ctx, string prefix)
+        {
+            if (prefix == null)
```

Comment:
Why do we need the prefix?
Wouldn't we simply continue to the next document? only after we consumed all counters of the previous one?

Reply:
When we have `commit = true` we are breaking from `GetCounters()` foreach loop.
then when we return to GetCounters loop we want to continue from were we left off. 
without the prefix we will start from the beginning again..